### ISSUE TITLE
contains solution with unit test: custom validator's returned 'error'(reason) string not consistently returned

### ISSUE THREAD

**Issue by llpage30** [ianstormtaylor/superstruct#99]
I have a data type 'one' that I want to return the value of the string if that value is not of length 1.
The StructError.reason, in these cases, should then be the string I returned.

The issue I am hitting and is eventually shown by the code below, is if I wrap my object definition containing this custom validator in an 'optional' call, the returned string 'gets lost' resulting in StructError.reason === undefined.  
This is a painful bug, and I cannot determine any work around for this.
Please fix this.

```javascript
const { superstruct } = require('superstruct');
const assert = require('assert');
// our 'one' type that should return the value in the structError.reason if value length != 1.
const TYPES = superstruct({
	types: {
		one: (value) => {
			if(value.length === 1) return true;
			return value;
		},
	},
});
const INNER = TYPES.partial({
	id: 'one',
	name: 'string',
});
// The use case I have is with a nested definition. These all work....
const OUTER_WORKS1 = TYPES.partial({
	no: 'one',
	user: INNER,
});
const OUTER_WORKS2 = TYPES.partial({
	no: 'one',
	user: [INNER],
});
// I wanted to have that nested definition be optional, but can't as the reason string does not come through.....
const OUTER_FAILS1 = TYPES.partial({
	no: 'one',
	user: TYPES.optional(INNER),
});
const OUTER_FAILS2 = TYPES.partial({
	no: 'one',
	user: TYPES.optional([INNER]),
});

// some data objects that will use in the asserts...
const innerDataGood = { id: '1', name: 'hello' };
const innerDataBad = { id: '11', name: 'hello' };
const outerDataGood = { no: 'y', user: innerDataGood };
const outerDataBad1 = { no: 'yes', user: innerDataGood };
const outerDataBad2 = { no: 'y', user: innerDataBad };
const outerData2Good = { no: 'y', user: [innerDataGood] };
const outerData2Bad1 = { no: 'yes', user: [innerDataGood] };
const outerData2Bad2 = { no: 'y', user: [innerDataBad] };

/**
  This all works in this function.
   This is here to indicate the structures above are valid and do return the reason string I wanted
   note: non of the FAILS cases are in this function.
**/
const baselineVerifyData = () => {
	let goodResult = INNER.validate(innerDataGood);
	assert.deepEqual(goodResult[1], innerDataGood);

	goodResult = OUTER_WORKS1.validate(outerDataGood);
	assert.deepEqual(goodResult[1], outerDataGood);

	goodResult = OUTER_WORKS2.validate(outerData2Good);
	assert.deepEqual(goodResult[1], outerData2Good);


	let badResult = INNER.validate(innerDataBad);
	assert.equal(badResult[0].reason, innerDataBad.id);

	badResult = OUTER_WORKS1.validate(outerDataBad1);
	assert.equal(badResult[0].reason, outerDataBad1.no);

	badResult = OUTER_WORKS1.validate(outerDataBad2);
	assert.equal(badResult[0].reason, innerDataBad.id);

	badResult = OUTER_WORKS2.validate(outerData2Bad1);
	assert.equal(badResult[0].reason, outerData2Bad1.no);

	badResult = OUTER_WORKS2.validate(outerData2Bad2);
	assert.equal(badResult[0].reason, innerDataBad.id);

};

/**
 Here is the showing of the issue I am reporting
the assert.notEqual() shows the issue. in a correct world this should 
have succeeded when it was assert.equal(), but here because of the issue
I need to use notEqual
***/
const failingData = () => {
	let goodResult = OUTER_FAILS1.validate(outerDataGood);
	assert.deepEqual(goodResult[1], outerDataGood);

	goodResult = OUTER_FAILS2.validate(outerData2Good);
	assert.deepEqual(goodResult[1], outerData2Good);


	let badResult = OUTER_FAILS1.validate(outerDataBad1);
	assert.equal(badResult[0].reason, outerDataBad1.no);

	badResult = OUTER_FAILS1.validate(outerDataBad2);
	assert.notEqual(badResult[0].reason, innerDataBad.id);

	badResult = OUTER_FAILS2.validate(outerData2Bad1);
	assert.equal(badResult[0].reason, outerData2Bad1.no);

	badResult = OUTER_FAILS2.validate(outerData2Bad2);
	assert.notEqual(badResult[0].reason, innerDataBad.id);
};
```

**Comment by llpage30** [comment]
The bug resides in this code block: https://github.com/ianstormtaylor/superstruct/blob/master/src/kinds.js#L774-L786

A javascript unit test definition you could add to your test/fixtures/object-objects :
[nested-property-custom-validator.txt](https://github.com/ianstormtaylor/superstruct/files/1803437/nested-property-custom-validator.txt)
Will illustrate the issue.

Basically you are using 'union' for 'optional'. and the definition of optional is will match the def, or can be undefined... I'm thinking you may need to 'break out' the optional to be its own function instead of a call to union. and then it will return the 1st failure.

**Comment by llpage30** [comment]
FIXED kinds.js  with unit test
[kinds.txt](https://github.com/ianstormtaylor/superstruct/files/1803561/kinds.txt)
[nested-property-custom-validator.txt](https://github.com/ianstormtaylor/superstruct/files/1803565/nested-property-custom-validator.txt)

**Comment by Barterr** [comment]
I'm encountering the same problem, it my type is optional the reason string is always undefined.

@llpage30 Could you submit a pull request with the fix? Pretty hard to find what changed by looking at a .txt

**Comment by tomspeak** [comment]
Would also be interested in this fix, can we get a diff of the changes please?

**Comment by Barterr** [comment]
I added a pull request with the fix and a test #101

**Comment by ianstormtaylor** [comment]
Thanks @Barterr!
